src/server.rs

```diff
@@ -1,34 +1,48 @@
+use std::convert::Infallible;
+use std::error::Error;
+use std::future::Future;
 use std::sync::Arc;
 
-use hyper::rt::{self, Future};
-use hyper::service::service_fn;
+use hyper::service::{make_service_fn, service_fn};
 use hyper::{Body, Request, Response, Server};
+use tokio::runtime::Runtime;
+use tokio::spawn;
 
 const PRODUCTION_PORT: u16 = 80;
 const DEVELOPMENT_PORT: u16 = 8082;
 
-pub fn start_server<S, F>(development_mode: bool, state: Arc<S>, handler: fn(Arc<S>, Request<Body>) -> F)
-where
+pub fn start_server<S, E, F>(
+    development_mode: bool,
+    state: Arc<S>,
+    handler: fn(Arc<S>, Request<Body>) -> F,
+) where
     S: Send + Sync + 'static,
-    F: Future<Item = Response<Body>, Error = hyper::error::Error> + Send + 'static,
+    E: Error + Send + Sync + 'static,
+    F: Future<Output = Result<Response<Body>, E>> + Send + 'static,
 {
-    let port = if development_mode {
-        DEVELOPMENT_PORT
-    } else {
-        PRODUCTION_PORT
-    };
+    Runtime::new().expect("Only should be called from main").block_on(async {
+        let port = if development_mode {
+            DEVELOPMENT_PORT
+        } else {
+            PRODUCTION_PORT
+        };
 
-    let addr = ([0, 0, 0, 0], port).into();
-    info!("Spinning up server on {:?}", addr);
+        let addr = ([0, 0, 0, 0], port).into();
+        info!("Spinning up server on {:?}", addr);
 
-    let new_service = move || {
-        let copied_state = state.clone();
-        service_fn(move |req| handler(copied_state.clone(), req))
-    };
+        let new_service = make_service_fn(move |_| {
+            let copied_state = state.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req| handler(copied_state.clone(), req)))
+            }
+        });
```

Comment:
Little confused by what is going on here

Reply:
It's basically a ton of jank to pass in the state properly. Hyper has a notion of "services" that deal with requests.

Hyper wants a thing that creates our service using context. `make_service_fn` lets us use a function to create a service from context (context we don't need symbolized by the underscore). This needs to be async, thus the async block. We use `service_fn` to return a function as a service.

All the `move` nonsense is because we need to thread through a reference to the state, and we want it to take ownership of the `state`. However, since these functions can be run repeatedly, we need to `clone` the `Arc` at each stage, not just `move` it.

This is by far the most arcane bit of the code haha. We can talk more about this offline